### ISSUE TITLE
Do we have "pleasant" migrations ?

### ISSUE THREAD

**Issue by buxx** [tracim/tracim#1810]
This issue is a discussion thread which follow discussion with @PhilippeAccorsi .

Do we have "pleasant" migrations between Tracim versions ? If i have a 1.x version and i update code to 1.x+1 (or 1.x+y) what happen ?

* Nothing can crash, automated migration applied and Tracim start
* Nothing can crash,  automated migration applied and Tracim stop asking me to apply manual step if any
* Possible crash: configuration file (or docker ENV) may be updated manually
* Oh, one of these possibilities depending on the situation

Maybe we can start to enumerate what is technically possible and what is already done to ensure pleasant migrations.

**Comment by inkhey** [comment]
@buxx 
It depend what you mean about "migration".

First between 1.x to 2.x, there is huge difference, so you need to migrate 1.x to last 1.x version and then migrate to 2.X because we do not have database migration script of 1.X in 2.X.
migration of calendar is a really specific case : https://github.com/tracim/tracim/issues/1612.

Migration from 2.x to another 2.x is pleasant if you do apply migration script, (instead it crash), running setup_default_backend script is done in order to accomplished the first possibility:

- Nothing can crash, automated migration applied and Tracim start

That's the goal, but we should be vigilant with some config change like change making path mandatory, because there can make tracim don't start correctly because of change of default config.

Trouble with database is less likely to happened in my opinion has we have test running all migrations script in both side: `tests/migration/test_migration.py`

**Comment by buxx** [comment]
My issue concern is about Tracim v2 (sorry for mistake in topic ... I edit it)

> That's the goal, but we should be vigilant with some config change like change making path mandatory, because there can make tracim don't start correctly because of change of default config.

Okay, that's the point. Tracim configurations is always back-compatible ? If no, what happen ?

Do docker image of Tracim work like you say to ? If i simply change Tracim image from 2.1 to 2.2 and restart my container, am i sure of no crashs ?

**Comment by inkhey** [comment]
> My issue concern is about Tracim v2 (sorry for mistake in topic ... I edit it)
> 
> > That's the goal, but we should be vigilant with some config change like change making path mandatory, because there can make tracim don't start correctly because of change of default config.
> 
> Okay, that's the point. Tracim configurations is always back-compatible ? If no, what happen ?

I don't think we decide something clear about this. Tracim configuration should always be forward-compatible, but back-compatible is unclear has new version mean new parameters not handled by old version. 
We can have deprecation of parameters which mean in version 2.1, old parameter exist, in version 2.2 old parameter work but warning exist about a new parameter, in version 2.3 or later, old parameter is dropped. So, il you need to go back from 2.3 to 2.1 you may have trouble with thoses.

> Do docker image of Tracim work like you say to ? If i simply change Tracim image from 2.1 to 2.2 and restart my container, am i sure of no crashs ?

As we do not have test on explicitly this kind of things, it's difficult to be sure, but yes, the best way to go from 2.1 to 2.2 is doing this, if you do not use exotic configuration this should not crash. 

Currently i see a situation where exotic configuration make this untrue. If you decide explicitly to set up session, depot, radicale or other dir needed to a dir which is not save outside of container, for example, you decide to store session data in /tmp in container. in 2.2, this will work flawly as session does not contain anything who need to stay alive (not like database and depot), but if you go to 2.3, you will have error at launching because session data dir become mandatory path, so you need the path to exist before running tracim, and as this is not a default path, you should create it manually.

**Comment by inkhey** [comment]
@buxx in all case, even if migration crashed for configuration issue, there is no reason to have un-explicit crash. if you configuration is wrong or missing for some reason, we should be explicit about it.

**Comment by buxx** [comment]
Okay. For future, I suggest imagining process to validate working of our migrations.
